cadence/server/api_actions.go

```diff
@@ -121,85 +130,136 @@ func liquidsoapRequest(path string) (message string, err error) {
 	return message, nil
 }
 
-var preTitle, preArtist, nowTitle, nowArtist string = "-", "-", "-", "-"
-var preHost, preMountpoint, nowHost, nowMountpoint string = "-", "-", "-", "-"
-var preListeners, nowListeners float64 = -1, -1
+func liquidsoapSkip() (message string, err error) {
+	clog.Debug("liquidsoapRequest", "Connecting to liquidsoap service...")
+	conn, err := net.Dial("tcp", c.SourceAddress+c.SourcePort)
+	defer conn.Close()
+	if err != nil {
+		clog.Error("liquidsoapRequest", "Failed to connect to audio source server.", err)
+		return "", err
+	}
+	fmt.Fprintf(conn, "cadence1.skip\n")
+	// Listen for response
+	message, _ = bufio.NewReader(conn).ReadString('\n')
+	fmt.Fprintf(conn, "quit"+"\n")
+	return message, nil
+}
 
-// Takes no arguments.
-// Returns nothing, but sends updated stream info to SSE and sets the global variables.
-// It is launched as a goroutine by init.
+// Watches the music directory (CSERVER_MUSICDIR) for any changes, and reconfigures the database.
+func filesystemMonitor() {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		clog.Error("fileSystemMonitor", "Error creating watcher.", err)
+	}
+	defer watcher.Close()
+	err = watcher.Add(c.MusicDir)
+	if err != nil {
+		clog.Error("fileSystemMonitor", "Error adding music directory to watcher.", err)
+	}
+	done := make(chan bool)
+	go func() {
+		for {
+			select {
+			case _, ok := <-watcher.Events:
+				if !ok {
+					continue
+				}
+				clog.Info("fileSystemMonitor", "Change detected in music library.")
+				dbRefresh()
+			case err, ok := <-watcher.Errors:
+				if !ok {
+					continue
+				}
+				clog.Error("fileSystemMonitor", "Error watching music library.", err)
+			}
+		}
+	}()
+	<-done
+}
+
+// Watches the Icecast status page and updates stream info for SSE.
 func icecastMonitor() {
-	for {
-		// If the stream data get errors, we set default "blank" stream data and continue so Cadence API will not completely fail
-		resp, err := http.Get("http://" + c.StreamAddress + c.StreamPort + "/status-json.xsl")
-		if resp != nil {
+	var prev = RadioInfo{}
+	go func() {
+		for {
+			time.Sleep(1 * time.Second)
+			resp, err := http.Get("http://" + c.StreamAddress + c.StreamPort + "/status-json.xsl")
+			if resp != nil {
+				defer resp.Body.Close()
+			}
+			if err != nil {
+				clog.Error("icecastMonitor", "Unable to stream data from the Icecast service.", err)
+				icecastDataReset()
+				continue
+			}
 			defer resp.Body.Close()
```

Comment:
Is there a reason we do this twice if `resp != nil`?

Reply:
There is not. I just got rid of "check if `resp != nil` " pattern.